### PR TITLE
Add Onyx Nova 2 frontlight controller

### PR DESCRIPTION
**Pull request by Galunid**
When run it crashes with
`E KOReader: [NativeThread]: failed to run script: frontend/ui/widget/frontlightwidget.lua:426: bad argument #1 to 'floor' (number expected, got nil)`
Do I need to add anything to frontend part? I'm pretty sure it everything should be handled already
Pinging @zwim @pazos since you probably have more idea

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/android-luajit-launcher/248)

### CODE REVIEW

**Comment by pazos** [comment]
@Galunid: Onyx and Likebooks are open android. Onyx even has an SDK. I would like to try `System.putInt` before resorting on non-documented tricks (API wise), like writting to sysfs nodes directly.

**Comment by pazos** [comment]
> E KOReader: [NativeThread]: failed to run script: frontend/ui/widget/frontlightwidget.lua:426: bad argument #1 to 'floor' (number expected, got nil)

That error means a non updated branch on the frontend. In master you'll never trigger an error in `frontend/ui/widget/frontlightwidget.lua` because the native dialog implementation overrides that widget.

File: app/src/org/koreader/launcher/device/DeviceInfo.kt [hunk 1]
```diff
@@ -142,6 +145,13 @@ object DeviceInfo {
         CREMA_0650L = BRAND.contentEquals("crema") && PRODUCT.contentEquals("keplerb")
         deviceMap[EinkDevice.CREMA_0650L] = CREMA_0650L
 
+        // Onyx NOVA 2
+        ONYX_NOVA = (MANUFACTURER.contentEquals("onyx")
+                && PRODUCT.contentEquals("nova2")
+                && DEVICE.contentEquals("nova2"))
+        deviceMap[EinkDevice.ONYX_NOVA] = ONYX_NOVA
```
Comment (pazos) [inline app/src/org/koreader/launcher/device/DeviceInfo.kt:152]: This is used for supported e-ink updates. Is not needed for Onyx, as we rely on system/android driver.

**Comment by zwim** [comment]
Don't mix sysfs with getInt/putInt. This could be dangerous, when sysfs has different ranges than Android. (At least on Toino this is true.)

**Comment by zwim** [comment]
Don't mix sysfs with getInt/putInt. This could be dangerous, when sysfs has different ranges than Android. (At least on Toino this is true.)

**Comment by Galunid** [comment]
I'd say normally that's the case, but it seems that's exactly what onyx does
```
08-25 20:59:27.668   937   937 V SettingsProvider: Notifying for 0: content://settings/system/screen_warm_brightness
08-25 20:59:27.690   937   992 I CTMBrightnessController: ###setLightValue, path: /sys/class/backlight/warm/brightness, value: 120
```
The scales are the same. It doesn't need any permissions too, I was pretty surprised, but I could set the settings properties via adb without root

> > E KOReader: [NativeThread]: failed to run script: frontend/ui/widget/frontlightwidget.lua:426: bad argument #1 to 'floor' (number expected, got nil)
> 
> That error means a non updated branch on the frontend. In master you'll never trigger an error in `frontend/ui/widget/frontlightwidget.lua` because the native dialog implementation overrides that widget.

Could you explain a bit more? I did 
```
git checkout master
git pull upstream master
git checkout -b my-branch
git add platform/android/luajit-launcher
```
luajit-launcher is on my branch with additional commits. 
Not sure if I'm missing something else. As for `enableFrontlightSwitch` it works as a placeholder, since I had to implement it. I'll fix it once I get the basics to work.

**Comment by pazos** [comment]
@Galunid: did you rebase your branch?. Does your koreader branch contain https://github.com/koreader/koreader/blob/master/frontend/ui/elements/common_settings_menu_table.lua#L205-L210 ?

**Comment by Galunid** [comment]
@pazos I didn't rebase, I created a new branch based on today's master. Yes, it contains the part you mentioned. I didn't do that yesterday, when I first build it. I'll try to clean build it again, hopefully it helps.

**Comment by pazos** [comment]
> @pazos I didn't rebase, I created a new branch based on today's master. Yes, it contains the part you mentioned. I didn't do that yesterday, when I first build it. I'll try to clean build it again, hopefully it helps.

In that case it is probably fresh bug or a regresion. I didn't look at the code in frontlightwidget, but https://github.com/koreader/koreader/blob/b5d33058761625111d176123121bcc881864a64e/frontend/device/devicelistener.lua#L229-L232 isn't right for android.

**Comment by Galunid** [comment]
@pazos Yup, you're right, I did it via gesture. I can open it successfully from menu :-)

**Comment by Galunid** [comment]
Doesn't work anyway, but it gives me a place to start ;)

**Comment by zwim** [comment]
> > @pazos I didn't rebase, I created a new branch based on today's master. Yes, it contains the part you mentioned. I didn't do that yesterday, when I first build it. I'll try to clean build it again, hopefully it helps.
> 
> In that case it is probably fresh bug or a regresion. I didn't look at the code in frontlightwidget, but https://github.com/koreader/koreader/blob/b5d33058761625111d176123121bcc881864a64e/frontend/device/devicelistener.lua#L229-L232 isn't right for android.

I can confirm that activating the frontlight dialog by gesture gives the error message `: [NativeThread]: failed to run script: frontend/ui/widget/frontlightwidget.lua:426: bad argument #1 to 'floor' (number expected, got nil)
`. This is a bug. 
@pazos will you fix it, or shall I do it?

**Comment by Galunid** [comment]
`System.putInt` indeed needs `WRITE_SETTINGS` permission, I got it to work properly using sysfs nodes. Is this acceptable? As for `onyxsdk`, I'm not even sure if it's accurate, since it's been 7 years since last commit and the company doesn't comply with GPL and doesn't publish sources sadly. On the bright side, this controller should work on most modern ONYX devices, although I cannot test it.

**Comment by zwim** [comment]
I would use sysfs on the getXxx methods too.

**Comment by pazos** [comment]
> `System.putInt` indeed needs `WRITE_SETTINGS` permission, I got it to work properly using sysfs nodes. Is this acceptable? 

Yeah, totally acceptable.


> As for `onyxsdk`, I'm not even sure if it's accurate, since it's been 7 years since last commit and the company doesn't comply with GPL and doesn't publish sources sadly. On the bright side, this controller should work on most modern ONYX devices, although I cannot test it.

They release the sdk as gradle dependencies. You can check the demo in https://github.com/onyx-intl/OnyxAndroidDemo.

The code there is just a platform wrapper to do the same things on all platforms onyx use on their devices. Feel free to play with it if you wish.

**Comment by Galunid** [comment]
There are a few things I'm occupied with recently, so sorry for the delay. I played with it a bit yesterday, but didn't manage to get frontlight to work properly. `FrontLightDemoActivity` doesn't mention any warmth settings, neither does their documentation, so I'll probably give up on implementing frontlight using OnyxSDK, since it needs `WRITE_SETTINGS` permission too. I'll try to clean up this pr tomorrow, so hopefully it'll be ready to merge.

**Review by pazos — APPROVED** [review]
@Galunid: LGTM, please do it when it's ready.

**Comment by Galunid** [comment]
@pazos Should be ready to go. I had one issue with `warmth` gestures not working, but it doesn't seem to be related to this PR, since incorrect value (0) is passed to `OnyxWarmthController:setWarmth`. If you think it's alright, you can merge (I don't have permissions in this repo).

**Comment by pazos** [comment]
> I had one issue with warmth gestures not working, but it doesn't seem to be related to this PR

This old PR has some info about Onyx: https://github.com/koreader/koreader/pull/5331, not sure if it is useful.

> If you think it's alright, you can merge

It looks alright, but I cannot test. LGTM. If you wish you can wait a few days before bumping this submodule in the frontend. The other way around (bumping soon and wait if other users report bugs) is fine for me too :)

**Pull request merged**